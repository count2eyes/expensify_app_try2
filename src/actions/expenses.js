import React from "react";
import uuid from "uuid";

export const addExpense = ({ description, amount, note, createdAt }) => {
  return {
    type: "ADD_EXPENSE",
    expense: {
      description: description || "",
      amount: amount || 0,
      note: note || "Notes!(not like you didn't notice)",
      createdAt: createdAt || 0,
      id: uuid()
    }
  };
};

export const removeExpense = id => {
  return {
    type: "REMOVE_EXPENSE",
    id
  };
};

export const editExpense = (id, { description, amount, note, createdAt }) => {
  return {
    type: "EDIT_EXPENSE",
    updates: { description, amount, note, createdAt },
    id
  };
};
