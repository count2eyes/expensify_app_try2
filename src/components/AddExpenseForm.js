import React from "react";
import { connect } from "react-redux";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import "react-dates/initialize";

export class AddExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      focused: false,
      description: props.expense ? props.expense.description : "",
      amount: props.expense ? props.expense.amount : "",
      note: props.expense ? props.expense.note : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      error: ""
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;

    this.setState(() => ({ description }));
  };

  onAmountChange = e => {
    const amount = e.target.value;
    if (/^\d*(\.\d{0,2})?$/.test(amount)) {
      this.setState(() => ({ amount }));
    } else return undefined;
  };

  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onDateChange = date =>
    this.setState({
      createdAt: date
    });

  onSubmit = e => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: "Please provide expense description & amount!"
      }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: this.state.amount,
        note: this.state.note,
        createdAt: moment(this.state.createdAt)
      });
    }
  };

  render() {
    return (
      <div>
        <form onSubmit={this.onSubmit}>
          {this.state.error ? <h4>{this.state.error}</h4> : null}
          <input
            type="text"
            onChange={this.onDescriptionChange}
            value={this.state.description}
            placeholder="Description"
          />
          <input
            type="number"
            onChange={this.onAmountChange}
            value={this.state.amount}
            placeholder="Amount"
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.focused}
            onFocusChange={({ focused }) =>
              this.setState({
                focused
              })
            }
            isOutsideRange={() => false}
            numberOfMonths={1}
            id="your_unique_id"
          />
          <textarea
            onChange={this.onNoteChange}
            value={this.state.note}
            placeholder="Your expense note (optional)"
          />
          <button type="submit">Submit</button>
        </form>
      </div>
    );
  }
}
export default connect(state => state)(AddExpenseForm);
