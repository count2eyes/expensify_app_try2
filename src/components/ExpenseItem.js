import React from "react";
import moment from "moment";
import { Link } from "react-router-dom";

const ExpenseItem = props => {
  const { description, amount, note, createdAt, id } = props.expense;
  return (
    <div>
      <Link to={`/edit/${id}`}>
        <h3>{description}</h3>
      </Link>
      <p>
        ${amount / 100} - {moment(createdAt).format("ddd DD MMM, YYYY")} -{" "}
        <small>{note}</small>
      </p>
    </div>
  );
};

export default ExpenseItem;
