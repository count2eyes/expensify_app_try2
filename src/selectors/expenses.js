import moment from "moment";

const selector = (expenses, filters) => {
  return expenses
    .filter(expense => {
      const matchedText = filters
        ? expense.description.includes(filters.text)
        : true;
      const matchedStartDate = filters.startDate
        ? moment(expense.createdAt).isSameOrAfter(filters.startDate, "day")
        : true;
      const matchedEndDate = filters.endDate
        ? moment(expense.createdAt).isSameOrBefore(filters.endDate, "day")
        : true;

      return matchedText && matchedStartDate && matchedEndDate;
    })
    .sort((a, b) => {
      if (filters.sortBy === "date") {
        return b.createdAt - a.createdAt;
      } else if (filters.sortBy === "amount") {
        return b.amount - a.amount;
      }
    });
};

export default selector;
