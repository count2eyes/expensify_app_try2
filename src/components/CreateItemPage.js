import React from "react";
import { connect } from "react-redux";
import AddExpenseForm from "./AddExpenseForm";
import { addExpense } from "../actions/expenses";

export class CreateItem extends React.Component {
  constructor(props) {
    super(props);
  }
  onSubmit = stateObj => {
    this.props.addExpense(stateObj);
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <h3>Add your expense here</h3>
        <AddExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  addExpense: expense => dispatch(addExpense(expense))
});

export default connect(
  undefined,
  mapDispatchToProps
)(CreateItem);
