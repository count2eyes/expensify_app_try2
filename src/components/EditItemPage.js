import React from "react";
import AddExpenseForm from "./AddExpenseForm";
import { editExpense, removeExpense } from "../actions/expenses";
import { connect } from "react-redux";

export class EditItem extends React.Component {
  constructor(props) {
    super(props);
  }

  onSubmit = stateObj => {
    this.props.editExpense(this.props.expense.id, stateObj);
    this.props.history.push("/");
  };

  onClick = () => {
    this.props.removeExpense(this.props.expense.id);
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <h3>Edit your expense</h3>
        <AddExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
        <button onClick={this.onClick}>Remove</button>
      </div>
    );
  }
}
const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      expense => expense.id === props.match.params.id
    )
  };
};

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, stateObj) => dispatch(editExpense(id, stateObj)),

  removeExpense: id => dispatch(removeExpense(id))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditItem);
