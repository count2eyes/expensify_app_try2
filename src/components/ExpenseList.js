import React from "react";
import selector from "../selectors/expenses";
import { connect } from "react-redux";
import ExpenseItem from "./ExpenseItem";

export class ExpenseList extends React.Component {
  render() {
    return (
      <div>
        {this.props.expenses.length === 0 ? (
          <h3>No expense</h3>
        ) : (
          selector(this.props.expenses, this.props.filters).map(expense => (
            <ExpenseItem expense={expense} key={expense.id} />
          ))
        )}
      </div>
    );
  }
}

export default connect(state => state)(ExpenseList);
