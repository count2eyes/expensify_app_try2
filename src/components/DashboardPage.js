import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";

const Dashboard = () => (
  <div>
    <h3>Your expenses</h3>
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default Dashboard;
