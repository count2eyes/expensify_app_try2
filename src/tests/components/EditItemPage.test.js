import React from "react";
import { shallow } from "enzyme";
import { EditItem } from "../../components/EditItemPage";
import expenses from "../fixtures/expenses";

let wrapper, editExpense, removeExpense, history;

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = {
    push: jest.fn()
  };
  wrapper = shallow(
    <EditItem
      editExpense={editExpense}
      removeExpense={removeExpense}
      history={history}
      expense={expenses[0]}
    />
  );
});

test("should render editItem page correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should submit edited item correctly", () => {
  wrapper.find("Connect(AddExpenseForm)").prop("onSubmit")(expenses[0]);
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
});

test("should remove item successfully", () => {
  wrapper.find("button").simulate("click");
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(removeExpense).toHaveBeenLastCalledWith(expenses[0].id);
});
