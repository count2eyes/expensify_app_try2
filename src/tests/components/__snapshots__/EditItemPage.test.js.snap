// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render editItem page correctly 1`] = `
<div>
  <h3>
    Edit your expense
  </h3>
  <Connect(AddExpenseForm)
    expense={
      Object {
        "amount": 2.75,
        "createdAt": "1970-01-01T00:00:00.000Z",
        "description": "coffee",
        "id": "1",
        "note": "decaff-rebuff",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
